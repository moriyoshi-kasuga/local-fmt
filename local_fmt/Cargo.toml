[package]
name = "local-fmt"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "This crate provides localizable format strings."
authors.workspace = true
repository.workspace = true
license.workspace = true
categories.workspace = true
keywords.workspace = true

[dependencies]
local-fmt-derive = { version = "^0", path = "../local_fmt_derive", optional = true }

[features]
default = []
derive = ["local-fmt-derive"]
selected = ["local-fmt-derive?/selected"]
global = ["local-fmt-derive?/global"]

[lib]
name = "local_fmt"

[lints]
workspace = true
