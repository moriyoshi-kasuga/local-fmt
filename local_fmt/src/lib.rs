use std::collections::HashMap;
use std::hash::Hash;

#[cfg(feature = "macros")]
pub use local_fmt_macros as macros;

pub mod message;
pub use message::*;

pub struct LocalFmt<L, M> {
    messages: HashMap<L, M>,
    global: fn() -> L,
}

impl<L: Eq + Hash, M> LocalFmt<L, M> {
    pub fn new(messages: HashMap<L, M>, global: fn() -> L) -> Self {
        Self { messages, global }
    }

    pub fn get_message(&self) -> &M {
        #[allow(clippy::expect_used)]
        self.messages
            .get(&(self.global)())
            .expect("message not found")
    }
}
