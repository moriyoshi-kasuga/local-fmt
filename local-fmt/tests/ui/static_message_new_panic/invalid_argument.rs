use local_fmt::{RefMessageFormat, StaticMessage};

const _: StaticMessage<1> = StaticMessage::<1>::new_panic(&[
    RefMessageFormat::RefText("Hello, world! "),
    RefMessageFormat::Placeholder(1),
]);

fn main() {}
