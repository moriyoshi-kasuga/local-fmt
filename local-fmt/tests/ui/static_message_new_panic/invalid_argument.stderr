error[E0080]: evaluation of constant value failed
 --> tests/ui/static_message_new_panic/invalid_argument.rs:3:29
  |
3 |   const _: StaticMessage<1> = StaticMessage::<1>::new_panic(&[
  |  _____________________________^
4 | |     RefMessageFormat::RefText("Hello, world! "),
5 | |     RefMessageFormat::Placeholder(1),
6 | | ]);
  | |__^ evaluation panicked: Invalid argument number: 1 is out of the allowed range (0 <= number < 1).
